src/main/kotlin/org/elaastic/questions/player/components/dashboard/DashboardModelFactory.kt

```diff
@@ -0,0 +1,232 @@
+                        sequenceMonitoringModel.phase2State,
+                        evaluationCountByUser
+                    ),
+                    sequenceMonitoringModel
+                )
+            )
+        }
+
+        sequenceMonitoringModel.setLearners(learners)
+
+        return DashboardModel(
+            sequence,
+            openedPane,
+            previousSequence?.id,
+            nextSequence?.id,
+            sequenceMonitoringModel
+        )
+    }
+
+    /**
+     * Get the LearnerStateOnPhase of the attendee on the Response phase.
+     *
+     * @param attendee the attendee
+     * @param attendeeResponses the list of responses of the attendee
+     * @param responsePhaseState the state of the response phase
+     * @return the state of the attendee on the response phase
+     * @see LearnerStateOnPhase
+     */
+    private fun getAttendeeStateOnResponsePhase(
+        attendee: LearnerAssignment,
+        attendeeResponses: List<Response>,
+        responsePhaseState: DashboardPhaseState
+    ): LearnerStateOnPhase {
+
+        val hasResponseForPhase: Boolean = attendeeResponses.count {
+            it.attempt == 1
+                    && it.interaction.interactionType == InteractionType.ResponseSubmission
+                    && it.learner == attendee.learner
+        } == 1
+
+        return if (responsePhaseState == DashboardPhaseState.NOT_STARTED) {
+            LearnerStateOnPhase.WAITING
+        } else if (hasResponseForPhase) {
+            LearnerStateOnPhase.ACTIVITY_TERMINATED
+        } else {
+            LearnerStateOnPhase.ACTIVITY_NOT_TERMINATED
+        }
+    }
+
+    /**
+     * Get the LearnerStateOnPhase of the attendee on the Evaluation phase.
+     *
+     * A learner has finished the phase if he has evaluated exactly the number
+     * of responses required.
+     *
+     * @param attendee the attendee
+     * @param sequence the sequence
+     * @param evaluationPhaseState the state of the evaluation phase
+     * @param evaluationCountByUser the number of evaluations made by each user
+     * @return the LearnerStateOnPhase base on the argument
+     * @see EvaluationSpecification
+     */
+    private fun getAttendeeStateOnEvaluationPhase(
+        attendee: LearnerAssignment,
+        sequence: Sequence,
+        evaluationPhaseState: DashboardPhaseState,
+        evaluationCountByUser: Map<LearnerAssignment, Long>
+    ): LearnerStateOnPhase {
+
+        // To not use any Service in a Factory, we need to pass the evaluationCountByUser as an argument
+        val nbEvaluationMade = evaluationCountByUser[attendee]
+
+        val hasMadeAllEvaluationForPhase: Boolean =
+            try {
+                sequence.getEvaluationSpecification().responseToEvaluateCount.toLong() == nbEvaluationMade
+            } catch (e: IllegalStateException) {
+                false /* If the sequence isn't initialized an Exception his throw by the getEvaluationSpecification function */
+            }
+
```

Comment:
This code must be extracted in a method so that it is called only if needed.

Reply:
Fix in 66b5059c77a8a4d3960e696eb38f0176b9028346

Reply:
Real fix in 555dbda9d568bc4571f2d5ece71d17e02ac53aed